[Gemma2] Enable Gemma2 Inference on Gaudi

Luca-Calabria opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->
This PR continues the work done to enable Gemma2 on Gaudi here https://github.com/huggingface/optimum-habana/pull/1280.

Single Card  
`python3 run_generation.py --model_name_or_path google/gemma-2-9b --batch_size 1 --use_kv_cache --max_new_tokens 100 --use_flash_attention --use_hpu_graphs --bf16`

```
11/25/2024 11:49:11 - INFO - __main__ - device: hpu, n_hpu: 0, bf16: True
11/25/2024 11:49:11 - INFO - __main__ - Model initialization took 12.165s
11/25/2024 11:49:11 - INFO - __main__ - Graph compilation...
Warming up iteration 1/3
11/25/2024 11:49:12 - INFO - __main__ - Time to first token = 424.24080099954153ms
Warming up iteration 2/3
11/25/2024 11:49:14 - INFO - __main__ - Time to first token = 18.381826997938333ms
Warming up iteration 3/3
11/25/2024 11:49:15 - INFO - __main__ - Time to first token = 14.948340995033504ms
11/25/2024 11:49:15 - INFO - __main__ - Running generate...
11/25/2024 11:49:16 - INFO - __main__ - Time to first token = 15.03746800153749ms
11/25/2024 11:49:17 - INFO - __main__ - Time to first token = 14.983528999437112ms
11/25/2024 11:49:18 - INFO - __main__ - Time to first token = 15.013975997135276ms
11/25/2024 11:49:19 - INFO - __main__ - Time to first token = 14.970359003200429ms
11/25/2024 11:49:20 - INFO - __main__ - Time to first token = 15.033622999908403ms

Input/outputs:
input 1: ('DeepSpeed is a machine learning framework',)
output 1: ('DeepSpeed is a machine learning framework that enables training of large-scale deep learning models on a single GPU or across multiple GPUs. It is designed to be easy to use and highly scalable, making it a powerful tool for researchers and practitioners working with large-scale deep learning models.\n\nDeepSpeed is built on top of PyTorch, a popular deep learning framework, and provides a set of tools and libraries that make it easy to train large-scale models. It includes features such as zero-shot inference, which allows models to be',)


Stats:
---------------------------------------------------------------------------------
Input tokens
Throughput (including tokenization) = 92.2688307826808 tokens/second
Memory allocated                    = 19.41 GB
Max memory allocated                = 19.42 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 3.9148149259999627 seconds
---------------------------------------------------------------------------------
```

Multiple Cards with DeepSpeed
`python3 ../gaudi_spawn.py --use_deepspeed --world_size 8 run_generation.py --model_name_or_path google/gemma-2-9b --batch_size 1 --use_kv_cache --max_new_tokens 100 --use_flash_attention --use_hpu_graphs`

```
11/25/2024 17:25:15 - INFO - __main__ - device: hpu, n_hpu: 8, bf16: True
11/25/2024 17:25:15 - INFO - __main__ - Model initialization took 42.190s
11/25/2024 17:25:15 - INFO - __main__ - Graph compilation...
Warming up iteration 1/3
11/25/2024 17:25:17 - INFO - __main__ - Time to first token = 1137.7861259970814ms
Warming up iteration 2/3
11/25/2024 17:25:18 - INFO - __main__ - Time to first token = 19.362977996934205ms
Warming up iteration 3/3
11/25/2024 17:25:19 - INFO - __main__ - Time to first token = 13.938947005954105ms
11/25/2024 17:25:19 - INFO - __main__ - Running generate...
11/25/2024 17:25:20 - INFO - __main__ - Time to first token = 13.15586299460847ms
11/25/2024 17:25:21 - INFO - __main__ - Time to first token = 12.668373994529247ms
11/25/2024 17:25:22 - INFO - __main__ - Time to first token = 12.446482003724668ms
11/25/2024 17:25:23 - INFO - __main__ - Time to first token = 11.886986998433713ms
11/25/2024 17:25:24 - INFO - __main__ - Time to first token = 14.2978059957386ms

Input/outputs:
input 1: ('DeepSpeed is a machine learning framework',)
output 1: ('DeepSpeed is a machine learning framework that enables training of large-scale deep learning models on a single GPU or across multiple GPUs. It is designed to be easy to use and highly scalable, making it a popular choice for training large-scale models such as GPT-3 and BERT.\n\nDeepSpeed is built on top of PyTorch, a popular deep learning framework, and provides a set of tools and libraries that make it easy to train large-scale models. It includes features such as zero-shot learning, which allows models to',)


Stats:
-----------------------------------------------------------------------------------
Input tokens
Throughput (including tokenization) = 108.78837579390004 tokens/second
Memory allocated                    = 5.95 GB
Max memory allocated                = 5.97 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 4.5140280299965525 seconds
-----------------------------------------------------------------------------------
```

Luca-Calabria commented:
@libinta @emascarenhas @nedo99 this is a PR to replace the original https://github.com/huggingface/optimum-habana/pull/1280 with fixes on style

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_1504). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

regisss commented:
@Luca-Calabria It seems there is still an issue with the code format.

Can you also add a test [here](https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py) adding the model output [there](https://github.com/huggingface/optimum-habana/blob/a69c201df2b571fbd357fb43149aeb31a51a93b5/tests/test_text_generation_example.py#L102) please?

Luca-Calabria commented:
> @Luca-Calabria It seems there is still an issue with the code format.
> 
> Can you also add a test [here](https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py) adding the model output [there](https://github.com/huggingface/optimum-habana/blob/a69c201df2b571fbd357fb43149aeb31a51a93b5/tests/test_text_generation_example.py#L102) please?

@regiss I fixed the format issue and added test case for Gemma2

emascarenhas commented on Makefile:
```diff
@@ -22,12 +22,12 @@ REAL_CLONE_URL = $(if $(CLONE_URL),$(CLONE_URL),$(DEFAULT_CLONE_URL))
 # Run code quality checks
 style_check: clean
-	pip install -U pip ruff
+	python.exe -m pip install -U pip ruff
```
Why is the python.exe required here?

Luca-Calabria replied:
my mistake sorry. I left it there after my local test on windows machine. I fix it

regisss commented:
@Luca-Calabria I just pushed a fix for the text generation test. Can you rebase your branch on main and also add the model output to check the model accuracy [here](https://github.com/huggingface/optimum-habana/blob/a69c201df2b571fbd357fb43149aeb31a51a93b5/tests/test_text_generation_example.py#L102) please?

Luca-Calabria commented:
> @Luca-Calabria I just pushed a fix for the text generation test. Can you rebase your branch on main and also add the model output to check the model accuracy [here](https://github.com/huggingface/optimum-habana/blob/a69c201df2b571fbd357fb43149aeb31a51a93b5/tests/test_text_generation_example.py#L102) please?

@regisss I rebased the PR and added the output check to test suite.

regisss commented:
@Luca-Calabria The code quality check probably failed because of a new version of `ruff`, can you run
```
pip install -U ruff
```
to make sure it's updated?

Luca-Calabria commented:
These are the issues found running make style. I think they came from one rebased step because 'modeling_starcoder2.py' is not something I modified in this PR
![image](https://github.com/user-attachments/assets/9cf4336e-b6eb-4f76-bdbd-0e7cf0349c5c)

Luca-Calabria commented:
@regisss please try again

jiminha commented:
@Luca-Calabria  I only see single configuration (batch1, output token100). Could you also clarify what other configurations have been run with performance data (You can provide more details in HS ticket)?   @libinta

Luca-Calabria commented:
> @Luca-Calabria I only see single configuration (batch1, output token100). Could you also clarify what other configurations have been run with performance data (You can provide more details in HS ticket)? @libinta

@jiminha I set up just a single configuration because I intend this test case as a non-regression check between new code versions, and this is also the "standard" configuration used for the others models. 
I think I didn't catch you question: Do you mean we should check other configs here? or you are asking what are the benchmarking setup for Gemma2 Inference?
